File: my-folder/0131-palindrome-partitioning/solution.py
class Solution:
    def partition(self, s: str) -> List[List[str]]:
        res = []
        partition = []

        def is_pali(s,i,j):
            while i < j :
                if s[i] != s[j]:
                    return False
                i, j = i + 1, j - 1
            return True

        def dfs(i):
            if (i >= len(s)):
                res.append(partition.copy())
                return
            
            for j in range(i, len(s)):
                if is_pali(s, i, j):
                    partition.append(s[i: j+1])
                    dfs(j + 1)
                    partition.pop()
        dfs(0)
        return res
        
        


#s.length could be 16
# s = "aab"
# [[a,a,b], [aa, b]]

        

File: my-folder/0131-palindrome-partitioning/solution.java
class Solution {
    List<List<String>> result;
    public List<List<String>> partition(String s) {
      result = new ArrayList<>();
      dfs(0, s, new ArrayList<>());
      return result;
    }

    private void dfs(int idx, String s, ArrayList<String> path){
        if(idx == s.length()){
            result.add(new ArrayList<>(path));
            return;
        }

        for(int curr = idx; curr < s.length(); curr++){
            if(isPalindrome(idx, curr, s)){
                path.add(s.substring(idx, curr + 1));
                dfs(curr + 1, s, path);
                path.remove(path.size() - 1);
            }
        }
    }

// check if each partition is a palindrome
    private boolean isPalindrome(int left, int right, String s){
        while(left < right){
            if(s.charAt(left++) != s.charAt(right--)){
                return false;
            }
        }
        return true;
    }
}
